src/mpi/coll/allreduce/allreduce_intra_recursive_multiplying.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+#include <math.h>
+
+/*
+ * Algorithm: Recursive Multiplying
+ *
+ * This algorithm is a generalization of the recursive doubling algorithm,
+ * and it has three stages. In the first stage, ranks above the nearest
+ * power of k less than or equal to comm_size collapse their data to the 
+ * lower ranks. The main stage proceeds with power-of-k ranks. In the main
+ * stage, ranks exchange data within groups of size k in rounds with 
+ * increasing distance (k, k^2, ...). Lastly, those in the main stage 
+ * disperse the result back to the excluded ranks. Setting k according 
+ * to the network hierarchy (e.g., the number of NICs in a node) can 
+ * improve performance.
+ */
+
+
+int MPIR_Allreduce_intra_recursive_multiplying(const void *sendbuf,
+                                            void *recvbuf,
+                                            MPI_Aint count,
+                                            MPI_Datatype datatype,
+                                            MPI_Op op,
+                                            MPIR_Comm * comm_ptr, 
+                                            const int k,
+                                            MPIR_Errflag_t errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    /* Ensure the op is commutative */
+    MPIR_Assert(MPIR_Op_is_commutative(op));
```

Comment:
The pre-stage is the only reason that this won't work with non-commutative ops, right?

Reply:
This is correct. As discussed in-person, I will update the pre-stage to work with non-commutative ops in a followup PR.

Reply:
How about only assert if not pofk?

Reply:
I don't quite understand what you mean by "only". We don't currently assert if not pofk

Reply:
My suggestion is to move the assertion after line 45 as -
```
if (comm_size > pofk) {
    /* Non-pofk only works when the op is commutative */
    MPIR_Assert(MPIR_Op_is_commutative(op)); 
}
```

Reply:
Ah got it sorry, that's a good idea will do!